internal/gatewayapi/servicecluster.go

```diff
@@ -0,0 +1,88 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package gatewayapi
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/ptr"
+
+	"github.com/envoyproxy/gateway/internal/envoygateway/config"
+	"github.com/envoyproxy/gateway/internal/gatewayapi/resource"
+	"github.com/envoyproxy/gateway/internal/ir"
+	"github.com/envoyproxy/gateway/internal/utils"
+)
+
+func (t *Translator) ProcessServiceCluster(acceptedGateways []*GatewayContext, resources *resource.Resources, xdsIR resource.XdsIRMap) {
+	for _, g := range acceptedGateways {
+		if g == nil || g.Gateway == nil {
+			continue
+		}
+
+		irKey := t.getIRKey(g.Gateway)
+
+		if xdsIR[irKey].ProxyServiceCluster != nil {
+			continue
+		}
+
+		var svcName string
+		var serviceClusterName string
+		if t.MergeGateways {
+			svcName = t.expectedServiceName(string(t.GatewayClassName))
+			serviceClusterName = string(t.GatewayClassName)
+
+		} else {
+			svcName = t.expectedServiceName(fmt.Sprintf("%s/%s", g.Namespace, g.Name))
+			serviceClusterName = fmt.Sprintf("%s/%s", g.Namespace, g.Name)
+		}
+
+		svc := resources.GetService(t.ControllerNamespace, svcName)
+		if svc == nil {
+			return
+		}
+
+		ds := t.processServiceClusterDestinationSetting(serviceClusterName, svc, resources)
+		ds.IPFamily = getServiceIPFamily(svc)
+		xdsIR[irKey].ProxyServiceCluster = &ir.ProxyServiceCluster{
+			Name:        serviceClusterName,
+			Destination: ds,
+		}
+
+		if t.MergeGateways {
+			return
+		}
+	}
+}
+
+// expectedServiceName returns expected Kubernetes Service name referring to the proxy Service Cluster.
+func (t *Translator) expectedServiceName(name string) string {
+	hashedName := utils.GetHashedName(name, 48)
+	return fmt.Sprintf("%s-%s", config.EnvoyPrefix, hashedName)
+}
+
+func (t *Translator) processServiceClusterDestinationSetting(
```

Comment:
can we reuse functons here ?
https://github.com/envoyproxy/gateway/blob/ceec66655765bea49635f951fa075796ca47796d/internal/gatewayapi/route.go#L1616